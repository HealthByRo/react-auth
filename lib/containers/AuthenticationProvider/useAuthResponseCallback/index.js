"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = useAuthResponseCallback;

var _react = require("react");

function useAuthResponseCallback(existingTokenData, existingUserData, setTokenData, setUserData) {
  return (0, _react.useCallback)(function (response) {
    var _response$data = response.data,
        tokenData = _response$data.tokenData,
        userData = _response$data.userData;
    setTokenData(tokenData);
    setUserData(userData);
  }, [existingTokenData, existingUserData]);
}