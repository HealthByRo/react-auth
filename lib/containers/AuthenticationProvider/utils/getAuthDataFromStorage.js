"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = getAuthDataFromStorage;

var _jsBase = require("js-base64");

var _constants = require("../constants");

var _getItemFromStorage = _interopRequireDefault(require("./getItemFromStorage"));

function getAuthDataFromStorage() {
  var localStorageItem = (0, _getItemFromStorage["default"])(_constants.AUTH_KEY);

  if (localStorageItem) {
    try {
      var decodedLocalStorageItem = _jsBase.Base64.decode(localStorageItem);

      return JSON.parse(decodedLocalStorageItem);
    } catch (e) {
      return null;
    }
  }

  return localStorageItem;
}