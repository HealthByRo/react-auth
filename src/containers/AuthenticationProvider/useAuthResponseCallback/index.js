import { useCallback } from 'react';

export default function useAuthResponseCallback(existingTokenData, existingUserData, setTokenData, setUserData) {
  return useCallback((response) => {
    const {
      data: {
        tokenData,
        userData,
      },
    } = response;

    setTokenData(tokenData);
    setUserData(userData);
  }, [existingTokenData, existingUserData]);
}
