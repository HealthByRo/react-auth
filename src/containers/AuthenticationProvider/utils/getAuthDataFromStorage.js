import { Base64 } from 'js-base64';
import { AUTH_KEY } from '../constants';
import getItemFromStorage from './getItemFromStorage';

export default function getAuthDataFromStorage() {
  const localStorageItem = getItemFromStorage(AUTH_KEY);

  if (localStorageItem) {
    try {
      const decodedLocalStorageItem = Base64.decode(localStorageItem);

      return JSON.parse(decodedLocalStorageItem);
    } catch (e) {
      return null;
    }
  }

  return localStorageItem;
}
