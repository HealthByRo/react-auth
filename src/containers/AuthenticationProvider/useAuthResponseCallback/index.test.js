import React, { useState } from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';
import useAuthResponseCallback from '.';
import getAuthDataFromStorage from '../utils/getAuthDataFromStorage';

let response;

let authResponseProcessorInstances;

function TestComponent() {
  const [something, setSomething] = useState(1);
  const [tokenData, setTokenData] = useState();
  const [userData, setUserData] = useState();
  const authResponseProcessor = useAuthResponseCallback(tokenData, userData, setTokenData, setUserData);

  authResponseProcessorInstances.add(authResponseProcessor);

  return (
    <>
      <section
        key={1}
        data-testid="tokenDataId"
      >
        {JSON.stringify(tokenData)}
      </section>

      <section
        key={2}
        data-testid="userDataId"
      >
        {JSON.stringify(userData)}
      </section>

      <section
        key={3}
        data-testid="somethingDataId"
      >
        {something}
      </section>

      <button
        type="button"
        onClick={() => authResponseProcessor(response)}
      >
        Set response data to authResponseProcessor
      </button>

      <button
        type="button"
        onClick={() => setSomething(something + 1)}
      >
        Something
      </button>
    </>
  );
}

describe('processAuthResponse', () => {
  let container;

  beforeEach(() => {
    authResponseProcessorInstances = new Set();
    container = render(<TestComponent />);
  });

  describe('when authResponseProcessor is not called', () => {
    it('does not render anything inside section with test-dataid "tokenDataId"', () => {
      expect(container.getByTestId('tokenDataId').firstChild).toBeNull();
    });

    it('does not render anything inside section with test-dataid "userDataId"', () => {
      expect(container.getByTestId('userDataId').firstChild).toBeNull();
    });

    it('auth data saved in localStorage is not defined', () => {
      expect(getAuthDataFromStorage()).toBeUndefined();
    });

    it('creates one instance of authResponseProcessor even component has been rerender', () => {
      const button = container.queryByText('Something');

      fireEvent.click(button);

      expect(authResponseProcessorInstances.size).toEqual(1);
    });
  });

  describe('when clicks on "Set response data to authResponseProcessor" button which invoke authResponseProcessor(response)', () => {
    beforeEach(() => {
      response = {
        data: {
          tokenData: {
            key: 'TEST_TOKEN',
            status: 'TEST_STATUS',
            expiredAt: 'TEST_DATE_TIME',
          },
          userData: {
            id: 'TEST_USER_ID',
            email: 'TEST_USER_EMAIL',
          },
        },
      };

      const button = container.queryByText('Set response data to authResponseProcessor');

      fireEvent.click(button);
    });

    it('creates second instance of authResponseProcessor', () => {
      const button = container.queryByText('Something');

      fireEvent.click(button);

      expect(authResponseProcessorInstances.size).toEqual(2);
    });

    it('renders serialized tokenData from response inside section with test-dataid "tokenDataId"', () => {
      expect(container.getByTestId('tokenDataId').firstChild).not.toBeNull();
      expect(container.getByTestId('tokenDataId').firstChild).toMatchSnapshot();
    });

    it('renders serialized userData from response inside section with test-dataid "userDataId"', () => {
      expect(container.getByTestId('userDataId').firstChild).not.toBeNull();
      expect(container.getByTestId('userDataId').firstChild).toMatchSnapshot();
    });
  });
});
