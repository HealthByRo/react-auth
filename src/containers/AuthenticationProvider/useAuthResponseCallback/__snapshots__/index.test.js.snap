// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processAuthResponse when clicks on "Set response data to authResponseProcessor" button which invoke authResponseProcessor(response) renders serialized tokenData from response inside section with data-testid "tokenDataId" 1`] = `{"key":"TEST_TOKEN","status":"TEST_STATUS","expiredAt":"TEST_DATE_TIME"}`;

exports[`processAuthResponse when clicks on "Set response data to authResponseProcessor" button which invoke authResponseProcessor(response) renders serialized userData from response inside section with data-testid "userDataId" 1`] = `{"id":"TEST_USER_ID","email":"TEST_USER_EMAIL"}`;
